server/map_rpc_server.go

```diff
@@ -354,66 +354,98 @@ func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapL
 
 	var newRoot *trillian.SignedMapRoot
 	err = t.registry.MapStorage.ReadWriteTransaction(ctx, tree, func(ctx context.Context, tx storage.MapTreeTX) error {
-		writeRev, err := tx.WriteRevision(ctx)
+		writeRev, err := t.writeRevision(ctx, tree, tx, req.Revision)
 		if err != nil {
 			return err
 		}
-		if rev := req.Revision; rev != 0 && writeRev != rev {
-			return status.Errorf(codes.FailedPrecondition, "can't write to revision %v", rev)
+
+		hkv, err := t.writeLeaves(ctx, tree, hasher, tx, req.Leaves, writeRev)
+		if err != nil {
+			return err
 		}
-		glog.V(2).Infof("%v: Writing at revision %v", mapID, writeRev)
-		txRunner := t.newTXRunner(tree, tx)
-		smtWriter, err := merkle.NewSparseMerkleTreeWriter(ctx, req.MapId, writeRev, hasher, txRunner)
+
+		newRoot, err = t.updateTree(ctx, tree, hasher, tx, hkv, req.Metadata, writeRev)
 		if err != nil {
 			return err
 		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &trillian.SetMapLeavesResponse{MapRoot: newRoot}, nil
+}
 
-		hkv := make([]merkle.HashKeyValue, 0, len(req.Leaves))
-		for _, l := range req.Leaves {
-			if err := checkIndexSize(l.Index, hasher); err != nil {
-				return err
-			}
-			l.LeafHash = hasher.HashLeaf(mapID, l.Index, l.LeafValue)
+// writeRevision ensures the desired write revision is available and claim a lock on this revision
+// in this transaction so that competing writes transactinos for the next revision cannot all be
+// committed.
+func (t *TrillianMapServer) writeRevision(ctx context.Context, tree *trillian.Tree, tx storage.MapTreeTX, rev int64) (int64, error) {
+	writeRev, err := tx.WriteRevision(ctx)
+	if err != nil {
+		return 0, err
+	}
+	if rev != 0 && writeRev != rev {
+		return 0, status.Errorf(codes.FailedPrecondition, "can't write to revision %v", rev)
+	}
+	glog.V(2).Infof("%v: Writing at revision %v", tree.TreeId, writeRev)
```

Comment:
nit: I'd put the logging up the stack, and leave `writeRevision` a "pure" function.